vllm/entrypoints/openai/serving_chat.py

```diff
@@ -1129,31 +1184,33 @@ async def chat_completion_full_generator(
                 logprobs = None
 
             if self.use_harmony:
-                reasoning_content, final_content, is_tool_call = (
-                    parse_chat_output(token_ids))
-                if not request.include_reasoning:
-                    reasoning_content = None
-
-                if is_tool_call:
-                    # TODO(woosuk): Implement tool call for gpt-oss.
-                    # For now, only Responses API supports tool call for
-                    # gpt-oss.
-                    raise NotImplementedError(
-                        "Tool call in Chat Completion API is not supported "
-                        "for gpt-oss yet. Please use Responses API instead.")
-                else:
-                    # Normal message
-                    message = ChatMessage(
-                        role=role,
-                        reasoning_content=reasoning_content,
-                        content=final_content,
-                    )
+                if TYPE_CHECKING:
+                    assert self.tool_parser is not None
+                    assert self.reasoning_parser is not None
+                tool_parser = self.tool_parser(tokenizer)
+                reasoning_parser = self.reasoning_parser(tokenizer)
+                # NOTE: We use token_ids for openai tool parser
+                tool_call_info = tool_parser.extract_tool_calls(
+                    "", request=request, token_ids=token_ids)
+                reasoning_content, content = None, tool_call_info.content
+                if request.include_reasoning:
+                    reasoning_content, content = reasoning_parser.extract_reasoning_content(  # noqa: E501
+                        "",
+                        request=request,
+                        token_ids=token_ids)
```

Comment:
given that gpt-oss already uses a special path, what about keeping the original tool parser and reasoning parser interface and parse gpt-oss directly in the special path? This token_ids can be confusing for people implementing new parsers.

Reply:
Adding a `token_ids` field to the `reasoning_parser` would be a great improvement.
Features like this one [https://github.com/vllm-project/vllm/pull/18067](https://github.com/vllm-project/vllm/pull/18067) require `token_ids`.